Fix bugs in db_create_coverage_table() + new is_used_for_line_up procedure

Aeilert opened the pull request:
Hi @tonyfujs 

Here is a PR for the new `is_used_for_line_up` column and procedure + several fixes for the coverage and censoring calculations. 

For the coverage calculation the following important things have changed: 

* Add SUR as a special country
* Merge with cl_table to ensure that region_code is available for all countries
* Merge with incgrp table to calculate LIC/LMIC coverage

Note that I kept the `*_table()` name for the function and filename for the censoring and coverage functions, even though they now return lists. Partly because we have started moving files to the ingestion repo. But we could change that if you want. 

For the refyear table note the following: 

* `db_create_ref_year_table()` now filters surveys based on the `is_used_for_line_up` column. That means that the predicted mean is only calculated for 1) "normal" national surveys 2) U/R surveys when the "domain" is non-national and 3) urban _or_ rural surveys when there is no national survey  at all for a specific country
* For the rule to create `is_used_for_line_up`  see changes in `db_create_dsm_table()` 
*  `is_used_for_line_up` might not be a perfect name in case it is confused with selection of INC/CON surveys for the line up (but maybe it is good enough) 
* Removed pfw_table from `db_create_ref_year_table()` inputs (no longer used) 


**EDIT:** After taking  a look at the API code and thinking more about this: `is_used_for_line_up` and `is_used_for_aggregation` measure conceptually different things. I thus added a (modified) version of   `is_used_for_aggregation`  back into the output tables. I believe the only surveys that should be aggregated are the ones split by U/R domain (IND, CHN, IDN).  We need to take care that ARG is included in the lineup (meaning `is_used_for_line_up = TRUE` ) now does, but there is no need for any aggregation (`is_used_for_aggregation = FALSE`).

Commit:
Fix bugs in db_create_coverage_table()

* Add SUR as a special country
* Merge with cl_table to ensure that region_code is available for all countries
* Other minor adjustments

Commit:
Add LIC/LMIC rule to coverage/censoring

Add rule to censor WLD years based on coverage in LIC/LIMC countries.

Commit:
Fix handling of U/R surveys in line-up

* Make sure PCN U/R rule for line-up is correctly implemented
* Rename is_used_for_aggregation to is_used_for_line_up

See PIP-Technical-Team/pip_ingestion_pipeline#/66
Also relevant for pipdm#116, pipdm#113

Commit:
Remove pfw_table from db_create_ref_year_table()

pfw_table is no longer used in the code, but has remained as a parameter from previous versions

Commit:
More fixes for coverage

* Fix merges between pop_table, cl_table and incgrp_table
* Impute incgrp_hist 1981-86 based on 1987 value

Aeilert commented:
In case you want to review the complete flow here, you can use the following script. 
The income group table is pending a PR review in pipaux, but I have attached a version that you can use for now. 

[income_groups.zip](https://github.com/PIP-Technical-Team/pipdm/files/8041022/income_groups.zip)


```r
# Load AUX data
gdp_table <- pipaux::load_aux("gdp") %>% setDT()
pce_table <- pipaux::load_aux("pce") %>% setDT()
pop_table <- pipaux::load_aux("pop") %>% setDT()
ppp_table <- pipaux::load_aux("ppp") %>% setDT()
cpi_table <- pipaux::load_aux("cpi") %>% setDT()
cl_table <- pipaux::load_aux("country_list") %>% setDT()
censored <- pipaux::load_aux("censoring")

# Load incgrp table (new AUX file not implemented yet)
incgrp_table <- readRDS('../pipaux/tmp/income_groups.rds') %>% setDT()

# Load LCU table
f <- "Y:/pip_ingestion_pipeline/pc_data/_targets/objects/svy_mean_lcu_table"
lcu_table <- qs::qread(f)

# Fix ARG 1980 which is incorrectly coded
lcu_table[country_code == "ARG",]$pop_data_level <- "urban"
lcu_table[country_code == "ARG",]$reporting_level <- "urban"

# Create DSM table
dsm_table <- db_create_dsm_table(lcu_table = lcu_table,
                                 cpi_table = cpi_table,
                                 ppp_table = ppp_table)

# Create ref year table
ref_year_table <- db_create_ref_year_table(
  dsm_table = dsm_table,
  gdp_table = gdp_table,
  pce_table = pce_table,
  pop_table = pop_table,
  ref_years = 1981:2021,
  pip_years = 1977:2019
)

# Fix NRU region code
# Inconsistency between PFW and PCN masterfile
ref_year_table[country_code == "NRU",]$pcn_region_code <- "EAP"

# Fix ARG 1980 which is incorrectly coded
ref_year_table[country_code == "ARG",]$pop_data_level <- "urban"
ref_year_table[country_code == "ARG",]$reporting_level <- "urban"

# Create coverage list of tables
coverage_list <- db_create_coverage_table(
  ref_year_table = ref_year_table,
  incgrp_table = incgrp_table,
  cl_table = cl_table,
  pop_table = pop_table,
  ref_years = 1981:2021
)

# Create regional censoring table
db_create_censoring_table(censored = censored,
                          coverage_list = coverage_list,
                          coverage_threshold = 50)
```

Commit:
Add is_used_for_aggregation

Add back a correct version of is_used_for_aggregation.

This column is TRUE for surveys split by U/R domain, but FALSE otherwise,
including national rows for such surveys.

tonyfujs commented on R/db_create_coverage_table.R:
```diff
@@ -1,21 +1,27 @@
                                      ref_years,
                                      digits = 2,
                                      special_countries =
-                                       c("ARG", "CHN", "IDN", "IND")) {
+                                       c("ARG", "CHN", "IDN", "IND", "SUR")) {
```
2 thougts on this:
1) It may be worth finding a more explicit name. These countries could be special in many ways. Something like `countries_to_include` maybe?
2) We should start removing hard coded values from our codebase, and move it to a dedicated file in the pipeline.

Aeilert replied:
I agree. These are basically the countries where U/R surveys are used for the line-up and coverage should thus be based on U/R population numbers. Maybe something like `urban_rural_countries` , `non_national_pop_countries` ? 

I also agree with 2) but we don't really have a good system in place for that. I suggest keeping such improvements separate from the current PR.

Aeilert replied:
Renamed to `urban_rural_countries`. Let me know what you think

tonyfujs commented on R/db_create_coverage_table.R:
```diff
@@ -111,14 +133,33 @@ db_create_coverage_table <- function(ref_year_table,
+    dplyr::summarise(coverage = stats::weighted.mean(coverage, pop)) %>%
+    dplyr::group_by(reporting_year) %>%
+    dplyr::summarise(coverage = mean(coverage)) %>%
+    dplyr::mutate(incgroup_historical = "LIC/LMIC") %>%
```
What is this for? Just worried about hard coded values...

Aeilert replied:
I don't think there is any hardcoded values in these four lines of code? The reason we are calculating coverage by historical income groups is because the censoring of WLD year depends on coverage in LIC/LMIC countries, in addition global coverage (this is a "new" rule that was included in PCN in 2021).

tonyfujs commented on R/db_create_dsm_table.R:
```diff
@@ -122,20 +122,17 @@ db_create_dsm_table <- function(lcu_table,
-  dt[, # create number of rows per cache_id
-     n_rl := .N,
-     by = cache_id
-     ][,
-       # variable for aggregate
-       is_used_for_aggregation := fifelse(n_rl > 1, TRUE, FALSE)
-     ][,
-       # remove counter
-       n_rl := NULL
-     ]
+  dt[, is_used_for_aggregation := ifelse((dt$reporting_level %in% c("urban", "rural") & dt$n_rl == 2), TRUE, FALSE)]
+  dt$n_rl <- NULL
```
I feel that this code chunk (lines 127 to 132) should live in its own function and be unit tested. This is an important piece of the PIP logic, and it will be easier to manage and document in a self-contained function.

Aeilert replied:
Okay. I can try to add a function and corresponding test for this.

Aeilert replied:
Added. Let me know what you think

Aeilert commented:
> > * `is_used_for_line_up` might not be a perfect name in case it is confused with selection of INC/CON surveys for the line up (but maybe it is good enough)
> 
> Where is this one implemented? Maybe we should be more explicit about that part of the process...

See https://github.com/PIP-Technical-Team/pipdm/blob/master/R/db_select_lineup_surveys.R

Commit:
Address reviewer comments

* Seperate "is_used_for_line_up" logic into a seperate function w/ unit tests
* Rename "special_countries" parameter to "urban_rural_countries".

Commit:
Add country-year coverage as output

* Add a table with country-year-reporting_level coverage as output from db_create_coverage_table().
* Use TRUE/FALSE to summarize. No need for additional 1, 0 transformation.

Commit:
Change survey_year names

Rename survey_year and survey_year_2 to
survey_year_before and survey_year_after.

tonyfujs reviewed (approved):
Thanks @Aeilert 

Looks good to me. One comment though: I haven't been able to run the code you shared with me because of some issue with a pipload dependency. 
![image](https://user-images.githubusercontent.com/1552634/154979451-041cc498-35bb-40f0-ae0f-67d404e95cee.png)
